File: scripts/row4a-browser-proof.mjs
#!/usr/bin/env node

import fs from 'node:fs/promises';
import path from 'node:path';
import process from 'node:process';

function parseArgs(argv) {
  const out = {};
  for (let i = 0; i < argv.length; i += 1) {
    const key = argv[i];
    if (!key.startsWith('--')) continue;
    const value = argv[i + 1];
    if (!value || value.startsWith('--')) {
      throw new Error(`Missing value for ${key}`);
    }
    out[key.slice(2)] = value;
    i += 1;
  }
  return out;
}

function asText(value) {
  return String(value || '').trim();
}

function asList(value) {
  return Array.isArray(value) ? value.map((item) => asText(item)).filter(Boolean) : [];
}

async function writeJson(filePath, payload) {
  await fs.mkdir(path.dirname(filePath), { recursive: true });
  await fs.writeFile(filePath, `${JSON.stringify(payload, null, 2)}\n`, 'utf8');
}

function findExactDeskMembership(snapshot, expected) {
  const desk = Array.isArray(snapshot?.feed?.desk) ? snapshot.feed.desk : [];
  for (const row of desk) {
    if (!row || typeof row !== 'object') continue;
    const rowNoteId = asText(row.note_id || row.desk_note_id);
    const rowTitle = asText(row.title || row.title_pl);
    const rowMessageId = asText(row.source_message_id || row.message_id);
    const rowSignalIds = asList(row.source_signal_ids);
    const rowEngagementId = asText(row.engagement_id);
    const rowCaseId = asText(row.case_id);
    if (expected.messageId && rowMessageId !== expected.messageId) continue;
    if (expected.signalId && !rowSignalIds.includes(expected.signalId)) continue;
    if (expected.engagementId && rowEngagementId !== expected.engagementId) continue;
    if (expected.caseId && rowCaseId !== expected.caseId) continue;
    if (expected.noteId && rowNoteId !== expected.noteId) continue;
    return {
      found: true,
      card_id: rowNoteId,
      title: rowTitle,
      source_message_id: rowMessageId,
      source_signal_ids: rowSignalIds,
      engagement_id: rowEngagementId,
      case_id: rowCaseId,
    };
  }
  return {
    found: false,
    card_id: '',
    title: '',
    source_message_id: '',
    source_signal_ids: [],
    engagement_id: '',
    case_id: '',
  };
}

function expectedIdentityFromHandoffItem(item) {
  const traceId = asText(item.trace_id);
  return {
    messageId: asText(item.source_message_id || item.message_id),
    signalId: asText(item.signal_id),
    traceId: traceId || null,
    traceIdSource: traceId ? 'handoff.item.trace_id' : 'missing',
    engagementId: asText(item.engagement_id),
    caseId: asText(item.case_id),
    noteId: asText(item.note_id) || (asText(item.engagement_id) ? `desk-${asText(item.engagement_id)}` : ''),
    snapshotId: asText(item.snapshot_id),
    title: asText(item.title),
  };
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const proofDir = path.resolve(args['proof-dir'] || '');
  const handoffPath = path.resolve(args.handoff || '');
  const baseUrl = asText(args['base-url']);
  const login = asText(args.login);
  const password = asText(args.password);
  if (!proofDir || !handoffPath || !baseUrl || !login || !password) {
    throw new Error('Required args: --proof-dir --handoff --base-url --login --password');
  }

  const browserDir = path.join(proofDir, 'browser');
  const networkPath = path.join(browserDir, 'network.json');
  const consolePath = path.join(browserDir, 'console.json');
  const anchorsPath = path.join(browserDir, 'anchors.json');
  const exactMembershipPath = path.join(proofDir, 'exact-snapshot-membership.json');
  const beforeShot = path.join(browserDir, 'card-before-click.png');
  const detailShot = path.join(browserDir, 'detail-after-click.png');

  const handoff = JSON.parse(await fs.readFile(handoffPath, 'utf8'));
  if (handoff.actionable !== true || !handoff.item || typeof handoff.item !== 'object') {
    throw new Error('Handoff is not actionable.');
  }

  const item = handoff.item;
  const expected = expectedIdentityFromHandoffItem(item);

  const { firefox } = await import('playwright');
  const browser = await firefox.launch({ headless: true });
  const context = await browser.newContext({ ignoreHTTPSErrors: true });
  const page = await context.newPage();

  const network = [];
  const consoleMessages = [];
  const pageErrors = [];
  const failedRequests = [];

  page.on('console', (msg) => {
    consoleMessages.push({
      type: msg.type(),
      text: msg.text(),
    });
  });
  page.on('pageerror', (err) => {
    pageErrors.push({
      name: err.name,
      message: err.message,
    });
  });
  page.on('requestfailed', (request) => {
    failedRequests.push({
      url: request.url(),
      method: request.method(),
      failure: request.failure(),
    });
  });
  page.on('response', async (response) => {
    const url = response.url();
    const entry = {
      url,
      status: response.status(),
      method: response.request().method(),
    };
    if (
      url.includes('/wp-json/daszek/v3/operational-feed-snapshots/latest') ||
      url.includes('/wp-json/daszek/v3/engagements/')
    ) {
      try {
        entry.body = await response.text();
      } catch (_error) {
        entry.body = '';
      }
    }
    network.push(entry);
  });

  try {
    await page.goto(baseUrl, { waitUntil: 'domcontentloaded', timeout: 30000 });
    await page.locator('#login-form input[name="login"]').fill(login);
    await page.locator('#login-form input[name="password"]').fill(password);

    const latestPromise = page.waitForResponse(
      (response) =>
        response.url().includes('/wp-json/daszek/v3/operational-feed-snapshots/latest') &&
        response.request().method() === 'GET' &&
        response.status() === 200,
      { timeout: 30000 },
    );

    await page.locator('#login-form button[type="submit"]').click();
    await page.waitForFunction(
      () => {
        const main = document.getElementById('main-screen');
        return !!main && window.getComputedStyle(main).display !== 'none';
      },
      null,
      { timeout: 30000 },
    );

    const skipButton = page.locator('#onboarding-skip');
    if ((await skipButton.count()) > 0) {
      try {
        if (await skipButton.first().isVisible()) {
          await skipButton.first().click();
        }
      } catch (_error) {
        // Ignore if onboarding is not shown.
      }
    }

    const latestResponse = await latestPromise;
    const latestPayload = await latestResponse.json();
    const latestSnapshot = latestPayload && typeof latestPayload === 'object' ? latestPayload.snapshot || {} : {};

    const membership = findExactDeskMembership(latestSnapshot, expected);
    if (!membership.found) {
      throw new Error('Exact snapshot membership not found for Row4a card.');
    }
    if (expected.snapshotId && asText(latestSnapshot.snapshot_id) !== expected.snapshotId) {
      throw new Error(`Latest snapshot mismatch. expected=${expected.snapshotId} actual=${asText(latestSnapshot.snapshot_id)}`);
    }

    await writeJson(exactMembershipPath, {
      proof_dir: proofDir,
      handoff_snapshot_id: expected.snapshotId,
      latest_snapshot_id: asText(latestSnapshot.snapshot_id),
      latest_matches_handoff_snapshot: asText(latestSnapshot.snapshot_id) === expected.snapshotId,
      expected_message_id: expected.messageId,
      expected_signal_id: expected.signalId,
      expected_engagement_id: expected.engagementId,
      expected_case_id: expected.caseId,
      expected_note_id: expected.noteId,
      expected_title: expected.title,
      membership_found: membership.found,
      membership: membership,
      latest_source_run_id: asText(latestSnapshot?.source?.source_run_id),
      latest_trigger_message_id: asText(latestSnapshot?.source?.trigger_message_id),
    });

    const cardSelector = `button.record-main[data-open-note="${membership.card_id}"]`;
    await page.waitForSelector(cardSelector, { timeout: 30000 });
    await page.screenshot({ path: beforeShot, fullPage: true });

    const liveResponsePromise = page.waitForResponse(
      (response) =>
        response.url().includes(`/wp-json/daszek/v3/engagements/${encodeURIComponent(expected.engagementId)}/os-events`) &&
        response.request().method() === 'GET' &&
        response.status() === 200,
      { timeout: 30000 },
    );

    await page.locator(cardSelector).click();
    const liveResponse = await liveResponsePromise;
    const liveBody = await liveResponse.json();

    await page.waitForSelector('.detail-section-os-events .os-event-row', { timeout: 30000 });
    await page.screenshot({ path: detailShot, fullPage: true });

    const detailText = await page.locator('#detail-panel').innerText();
    const anchors = {
      proof_dir: proofDir,
      page_url: baseUrl,
      message_id: expected.messageId,
      signal_id: expected.signalId,
      trace_id: expected.traceId,
      trace_id_source: expected.traceIdSource,
      engagement_id: expected.engagementId,
      case_id: expected.caseId,
      note_id: membership.card_id,
      snapshot_id: expected.snapshotId,
      handoff_title: expected.title,
      latest_snapshot_id: asText(latestSnapshot.snapshot_id),
      latest_title: membership.title,
      latest_source_signal_ids: membership.source_signal_ids,
      latest_trigger_message_id: asText(latestSnapshot?.source?.trigger_message_id),
      live_request_url: liveResponse.url(),
      live_request_status: liveResponse.status(),
      live_response_ok: Boolean(liveBody && liveBody.ok),
      live_response_engagement_id: asText(liveBody && liveBody.engagement_id),
      live_response_items_count: Array.isArray(liveBody && liveBody.items) ? liveBody.items.length : 0,
      latest_matches_handoff_snapshot: asText(latestSnapshot.snapshot_id) === expected.snapshotId,
      latest_matches_signal: membership.source_signal_ids.includes(expected.signalId),
      latest_title_matches_handoff: membership.title === expected.title,
      detail_contains_latest_title: detailText.includes(membership.title),
      js_errors: pageErrors.length,
      failed_requests: failedRequests.length,
    };

    await writeJson(anchorsPath, anchors);
    await writeJson(networkPath, network);
    await writeJson(consolePath, {
      console: consoleMessages,
      pageErrors,
      requestsFailed: failedRequests,
    });
  } finally {
    await context.close();
    await browser.close();
  }
}

main().catch(async (error) => {
  console.error(error && error.stack ? error.stack : String(error));
  process.exit(1);
});

File: scripts/row4a_browser_proof.py
#!/usr/bin/env python
from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path
from typing import Any

from playwright.sync_api import sync_playwright

sys.path.insert(0, str(Path(__file__).resolve().parent))

from row4a_proof_support import (  # noqa: E402
    ENGAGEMENTS_PATH,
    LATEST_SNAPSHOT_PATH,
    ProofObservation,
    ProofPaths,
    ProofRecorder,
    ProofRequest,
    SnapshotMatch,
    as_text,
    expected_identity_from_handoff_item,
    find_exact_desk_membership,
    membership_payload,
    write_json,
)


TIMEOUT_MS = 30000
MAIN_SCREEN_VISIBLE_JS = """() => {
                    const main = document.getElementById('main-screen');
                    return !!main && window.getComputedStyle(main).display !== 'none';
                }"""


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--proof-dir", required=True)
    parser.add_argument("--handoff", required=True)
    parser.add_argument("--base-url", required=True)
    parser.add_argument("--login", required=True)
    parser.add_argument("--password", required=True)
    return parser.parse_args()


def expected_identity_from_handoff(handoff_path: Path) -> dict[str, str | None]:
    handoff = json.loads(handoff_path.read_text(encoding="utf-8"))
    if handoff.get("actionable") is not True or not isinstance(handoff.get("item"), dict):
        raise RuntimeError("Handoff is not actionable.")
    return expected_identity_from_handoff_item(handoff["item"])


def _is_latest_snapshot_response(response: Any) -> bool:
    return (
        LATEST_SNAPSHOT_PATH in response.url
        and response.request.method == "GET"
        and response.status == 200
    )


def _dismiss_onboarding(page: Any) -> None:
    skip_button = page.locator("#onboarding-skip")
    if skip_button.count() == 0:
        return
    try:
        if skip_button.first.is_visible():
            skip_button.first.click()
    except Exception:
        pass


def log_in_and_await_latest(page: Any, request: ProofRequest) -> Any:
    """Log in, wait for the operator screen, and return the awaited `latest` response."""
    page.goto(request.base_url, wait_until="domcontentloaded", timeout=TIMEOUT_MS)
    page.locator('#login-form input[name="login"]').fill(request.login)
    page.locator('#login-form input[name="password"]').fill(request.password)

    latest_response_info = page.expect_response(_is_latest_snapshot_response, timeout=TIMEOUT_MS)
    with latest_response_info as latest_wait:
        page.locator('#login-form button[type="submit"]').click()

    page.wait_for_function(MAIN_SCREEN_VISIBLE_JS, timeout=TIMEOUT_MS)
    _dismiss_onboarding(page)
    return latest_wait.value


def latest_snapshot_from(response: Any) -> dict[str, Any]:
    payload = response.json()
    return payload.get("snapshot", {}) if isinstance(payload, dict) else {}


def resolve_membership(latest_snapshot: dict[str, Any], expected: dict[str, Any]) -> dict[str, Any]:
    """Locate the Row4a card in the live snapshot, failing loudly on any identity mismatch."""
    membership = find_exact_desk_membership(latest_snapshot, expected)
    if not membership["found"]:
        raise RuntimeError("Exact snapshot membership not found for Row4a card.")
    actual_snapshot_id = as_text(latest_snapshot.get("snapshot_id"))
    if expected["snapshot_id"] and actual_snapshot_id != expected["snapshot_id"]:
        raise RuntimeError(
            f"Latest snapshot mismatch. expected={expected['snapshot_id']} actual={actual_snapshot_id}"
        )
    return membership


def open_card_and_read_detail(page: Any, paths: ProofPaths, engagement_id: str, card_id: str) -> tuple[Any, Any, str]:
    """Click the desk card and return (live os-events response, parsed body, detail text)."""
    card_selector = f'button.record-main[data-open-note="{card_id}"]'
    page.wait_for_selector(card_selector, timeout=TIMEOUT_MS)
    page.screenshot(path=str(paths.card_before_click), full_page=True)

    def is_os_events_response(response: Any) -> bool:
        return (
            f"{ENGAGEMENTS_PATH}{engagement_id}/os-events" in response.url
            and response.request.method == "GET"
            and response.status == 200
        )

    detail_response_info = page.expect_response(is_os_events_response, timeout=TIMEOUT_MS)
    with detail_response_info as detail_wait:
        page.locator(card_selector).click()

    live_response = detail_wait.value
    live_body = live_response.json()

    page.wait_for_selector(".detail-section-os-events .os-event-row", timeout=TIMEOUT_MS)
    page.screenshot(path=str(paths.detail_after_click), full_page=True)
    return live_response, live_body, page.locator("#detail-panel").inner_text()


def run_proof(page: Any, request: ProofRequest, recorder: ProofRecorder) -> int:
    paths = request.paths
    expected = request.expected
    latest_snapshot = latest_snapshot_from(log_in_and_await_latest(page, request))
    match = SnapshotMatch(
        proof_dir=paths.proof_dir,
        expected=expected,
        latest_snapshot=latest_snapshot,
        membership=resolve_membership(latest_snapshot, expected),
    )
    write_json(paths.exact_membership, membership_payload(match))

    live_response, live_body, detail_text = open_card_and_read_detail(
        page, paths, expected["engagement_id"], match.membership["card_id"]
    )
    recorder.record_anchors(
        ProofObservation(
            match=match,
            page_url=request.base_url,
            live_response=live_response,
            live_body=live_body,
            detail_text=detail_text,
        )
    )
    return 0 if recorder.critical_anchors_ok() else 1


def build_request(args: argparse.Namespace) -> ProofRequest:
    return ProofRequest(
        base_url=args.base_url,
        login=args.login,
        password=args.password,
        paths=ProofPaths.for_dir(Path(args.proof_dir).resolve()),
        expected=expected_identity_from_handoff(Path(args.handoff).resolve()),
    )


def main() -> int:
    request = build_request(parse_args())
    recorder = ProofRecorder(request.paths.proof_dir)

    with sync_playwright() as playwright:
        browser = playwright.firefox.launch(headless=True)
        context = browser.new_context(ignore_https_errors=True)
        page = context.new_page()
        recorder.attach(page)
        try:
            exit_code = run_proof(page, request, recorder)
        except Exception as exc:
            recorder.record_error(exc)
            exit_code = 1
        finally:
            recorder.persist(request.paths)
            context.close()
            browser.close()

    return exit_code


if __name__ == "__main__":
    raise SystemExit(main())
